tools/make_patch.c

```diff
@@ -0,0 +1,891 @@
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <errno.h>
+#include <string.h>
+#include <ctype.h>
+
+struct symbol {
+	struct symbol *next;
+	unsigned int value;
+	char name[];
+};
+
+struct patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct buffer {
+	size_t size;
+	char buffer[];
+};
+
+struct asmfile {
+	struct asmfile *next;
+	struct symbol *symbols;
+	char name[];
+};
+
+struct asmfile *asmfiles = NULL;
+
+void free_buffer(void *buf) {
+	struct buffer *buffer = (struct buffer *)((char *)buf - sizeof(struct buffer));
+	free(buffer);
+}
+
+void *create_buffer() {
+	struct buffer *buffer = malloc(sizeof(struct buffer) + 0x100);
+	if (!buffer) {
+		fprintf(stderr, "Error: Cannot allocate memory\n");
+		return NULL;
+	}
+
+	buffer->size = 0x100;
+	return buffer->buffer;
+}
+
+void *expand_buffer(void *buf, const size_t size) {
+	struct buffer *buffer = (struct buffer *)((char *)buf - sizeof(struct buffer));
+
+	if (size > buffer->size) {
+		buffer = realloc(buffer, (buffer->size += 0x100));
+		if (!buffer) {
+			free_buffer(buf);
+			fprintf(stderr, "Error: Cannot allocate memory\n");
+			return NULL;
+		}
+	}
+
+	return buffer->buffer;
+}
+
+void free_symbols(struct symbol *list) {
+	while (list) {
+		struct symbol *next = list->next;
+		free(list);
+		list = next;
+	}
+}
+
+void free_asmfiles(struct asmfile *list) {
+	while (list) {
+		if (list->symbols) free_symbols(list->symbols);
+		struct asmfile *next = list->next;
+		free(list);
+		list = next;
+	}
+}
+
+bool create_symbol(struct symbol **tip, const int value, const char *name) {
+	struct symbol *symbol;
+
+	symbol = malloc(sizeof(struct symbol) + strlen(name) + 1);
+	if (!symbol) {
+		fprintf(stderr, "Error: Cannot allocate memory\n");
+		return true;
+	}
+	symbol->value = value;
+	strcpy(symbol->name, name);
+
+	// Link it at the start of the rest of the list
+	symbol->next = *tip;
+	*tip = symbol;
+
+	return false;
+}
+
+const struct symbol *find_symbol(const struct symbol *symbols, const char *name) {
+	const struct symbol *symbol = symbols;
+	int namelen = strlen(name);
+	while (symbol) {
+		int symbolnamelen = strlen(symbol->name);
+		if (namelen > symbolnamelen) {
+			symbol = symbol->next; 
+			continue;
+		}
+		// If we're looking for a local label, only compare the part starting from the period
+		char *compare = (char *)symbol->name;
+		if (name[0] == '.') compare += symbolnamelen - namelen;
+		if (!strcmp(compare, name)) return symbol;
+		symbol = symbol->next;
+	}
+
+	return NULL;
+}
+
+int parse_address_bank(char *buffer_input) {
```

Comment:
Most of the work here is duplicated in `parse_address_address`. These should be united into a single function, `void parse_address(char *buffer_input, int &bank, int &address)`, which would take pointers to its two output values. Also it should not check for lengths of 2 and 4, since `001:004000` is technically valid. Instead it should just do `strtol` and then check the range of the actual integer values.

Reply:
Pass by references are not a thing in C... This can be fixed later in C++

Ignore lengths of 2 and 4, and just strtol and check range of int hex values.